### PR TITLE
EnsuringDiagnosticsHandlers uses weak references for non-trivial refs

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description
CosmosDiagnosticsHandler instances can be referenced by applications indirectly - via CosmosClientConfig - this is needed for example for dynamic reconfiguration. This can lead to unnecessary memory pressure when the application closes (and reopens) clients - but keeps references to the ClientTelemetryConfig instances. This would result in strong references from the diagnostics handlers to CosmosAsyncClient (and due to this netty connection pool buffers etc.). By switching the weak references in this PR the GC can clean-up the CosmsoAsyncClient and netty resources sooner.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientMetricsDiagnosticsHandler.java [hunk 1]
```diff
@@ -9,26 +9,33 @@
...
-        this.client = client;
+        this.clientRef =  new WeakReference<>(client);
     }
 
     @Override
     public void handleDiagnostics(CosmosDiagnosticsContext diagnosticsContext, Context traceContext) {
         checkNotNull(traceContext, "Argument 'traceContext' must not be null.");
 
+        CosmosAsyncClient clientSnapshot = this.clientRef.get();
+        if (clientSnapshot == null) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientMetricsDiagnosticsHandler.java:33]: when can this happen?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientTelemetryDiagnosticsHandler.java [hunk 2]
```diff
@@ -51,11 +53,17 @@ public void handleDiagnostics(CosmosDiagnosticsContext diagnosticsContext, Conte
 
         this.ensureInitialized();
 
+        ClientTelemetry clientTelemetrySnapshot = this.telemetryRef.get();
+        if (clientTelemetrySnapshot == null) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientTelemetryDiagnosticsHandler.java:57]: Same here? Is this condition possible? Is it intentional to just return from here? how will it affect the workflow of customers?

**Review by kushagraThapar — APPROVED** [review]
Change looks good to me, thanks @FabianMeiswinkel , just a couple of questions.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientTelemetryDiagnosticsHandler.java (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientTelemetryDiagnosticsHandler.java:57]: Yes, it is intentional. Only way to get here I can see are race conditions - remaining requests while client gets closed. Worst case we lose client-telemetry and/or metrics - nothing critical.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientMetricsDiagnosticsHandler.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientMetricsDiagnosticsHandler.java:33]: Only way to get here I can see are race conditions - remaining requests while client gets closed. Worst case we lose client-telemetry and/or metrics - nothing critical.

**Review by jeet1995 — APPROVED** [review]
LGTM - thanks!

**Review by lmolkova — APPROVED** [review]

**Pull request merged**